docs/developing-contracts/gear-program.md

```diff
@@ -5,25 +5,263 @@ sidebar_position: 2
 #![no_std]
 
-use gstd::msg;
+// External packages (crates) and internal modules import
+use codec::{Decode, Encode};
+use gstd::{debug, msg, prelude::*};
+use scale_info::TypeInfo;
+```
+
+```rust
+// Init function that is executed once upon contract initialization
+// Here is empty
+#[no_mangle]
+pub unsafe extern "C" fn init() {}
+
+// Handle function that processes the incoming message
+#[no_mangle]
+pub unsafe extern "C" fn handle() {}
+```
+
+It is the minimum necessary structure for our application to work. The init() function is executed once during the context initialization and can perform any actions once. The handle function is responsible for processing all incoming messages handled by our program.
+
+Next, we will add a Voting structure that will contain our main code for working with the program state.
+
+```rust
+#[derive(Clone)]
+pub struct State {
+    votes_received: BTreeMap<String, i32>,
+}
+
+impl State {
+    // Create a state
+    pub const fn new() -> Self {
+        Self {
+            votes_received: BTreeMap::new(),
+        }
+    }
 
+    // Add new candidate
+    pub fn add_candidate(&mut self, candidate: String) {
+        self.votes_received.insert(candidate, 0);
+    }
+
+    // Vote for the candidate by name. If candidate no exist add it
+    pub fn vote_for_candidate(&mut self, name: String) {
+        let counter = self.votes_received.entry(name).or_insert(0);
+        *counter += 1;
+    }
+}
+
+// The state itself (i.e. the variable state will be accessed through)
+static mut STATE: State = State::new();
+```
+
+We will also need to define the metadata structure for implementing the input/output communication interface. The method described is a binary map for mutual interaction between different programming languages. For example, since the program is compiled into WASM format, it only understands the byte language. To simplify the operation, we define data structures in advance for further encoding and decoding. For this, we use a special macro gstd::metadata!:
+
+```rust
+gstd::metadata! {
+   title: "Voting App",
+   handle:
+       input: Action,
+   state:
+       input: StateAction,
+       output: StateReply,
+}
+```
+
+Now let's start processing incoming messages. Every time our contract receives an incoming message, we will process it accordingly. Let's describe the handle() function:
+
+```rust
+#[derive(Debug, TypeInfo, Encode)]
+pub enum StateReply {
+   All(BTreeMap<String, i32>),
+   VotesFor(i32),
+}
+ 
+#[derive(Debug, TypeInfo, Decode)]
+pub enum StateAction {
+   All,
+   VotesFor(String),
+}
+
+// Handle function that processes the incoming message
 #[no_mangle]
 pub unsafe extern "C" fn handle() {
-    msg::reply(b"Hello world!", 0, 0);
+    let action: Action = msg::load().unwrap();
+
+    debug!("Received action: {:?}", action);
+
+    match action {
+        Action::AddCandidate(name) => {
+            STATE.add_candidate(name.clone());
+
+            msg::reply((), 0, 0);
```

Comment:
New `gstd` hasn't gas limit in `reply()` function.
```suggestion
            msg::reply((), 0);
```

Reply:
This tutorial was pushed by mistake